packages/infer-schema/src/to-neo4j-struct.ts

```diff
@@ -0,0 +1,151 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Session } from "neo4j-driver";
+import Node from "./classes/Node";
+import Property from "./classes/Property";
+import Relationship from "./classes/Relationship";
+import { Neo4jStruct, NodeMap, PropertyRecord, RelationshipMap } from "./types";
+import cleanTypeName from "./utils/clean-type-name";
+
+interface NodeTypePropertiesRecord extends PropertyRecord {
+    nodeType: string;
+    nodeLabels: string[];
+}
+interface RelationshipTypePropertiesRecord extends PropertyRecord {
+    relType: string;
+}
+
+type RelationshipRecord = {
+    from: string;
+    to: string;
+    relType: string;
+};
+
+export default async function toNeo4jStruct(sessionFactory: () => Session): Promise<Neo4jStruct> {
+    const nodes = await inferNodes(sessionFactory);
+    const relationships = await inferRelationships(sessionFactory);
+
+    return { nodes, relationships };
+}
+
+async function inferNodes(sessionFactory: () => Session): Promise<NodeMap> {
+    const nodes: NodeMap = {};
+    // Label properties
+    const session = sessionFactory();
+    const labelPropsRes = await session.readTransaction((tx) =>
+        tx.run(`CALL db.schema.nodeTypeProperties()
+    YIELD nodeType, nodeLabels, propertyName, propertyTypes, mandatory
+    RETURN *`)
+    );
+    await session.close();
+    if (!labelPropsRes?.records.length) {
+        return nodes;
+    }
+    const nodeTypeProperties = labelPropsRes.records.map((r) => r.toObject()) as NodeTypePropertiesRecord[];
+    new Set(nodeTypeProperties.map((nt) => nt.nodeType)).forEach((nodeType) => {
+        const propertiesRows = nodeTypeProperties.filter((nt) => nt.nodeType === nodeType);
+        if (!propertiesRows) {
+            return;
+        }
+        const { nodeLabels } = propertiesRows[0];
+        const node = new Node(nodeType, nodeLabels);
+        propertiesRows.forEach((p) => {
+            if (!p.propertyName) {
+                return;
+            }
+            node.addProperty(new Property(p.propertyName, p.propertyTypes, p.mandatory));
+        });
+        nodes[nodeType] = node;
+    });
+
+    return nodes;
+}
+
+async function inferRelationships(sessionFactory: () => Session): Promise<RelationshipMap> {
+    const relSession = sessionFactory();
+    const rels: RelationshipMap = {};
+
+    // Find all relationship types and their properties (if any)
+    const typePropsRes = await relSession.readTransaction((tx) =>
+        tx.run(`CALL db.schema.relTypeProperties()
+    YIELD relType, propertyName, propertyTypes, mandatory
+    RETURN *`)
+    );
+    await relSession.close();
+    const relTypePropertiesRecords = typePropsRes.records.map((r) =>
+        r.toObject()
+    ) as RelationshipTypePropertiesRecord[];
+    const uniqueRelTypes = new Set(relTypePropertiesRecords.map((nt) => nt.relType));
```

Comment:
I'm kinda curious about the decision to only take unique rel types here. When working on path filtering of search results in Bloom, I got tripped up by learning that there can be multiple relationships with the same type, but having different properties. It looks like the code below handles this by just picking the properties of the last one... ?

Reply:
FYI, we handled duplicate rel types in Bloom by also looking at the source/target node labels.

The code in the path segment filtering for search suggestions that handles that is here:
https://github.com/neo-technology/neo4j-insight/blob/6605410f0901eb7210dada6cdb73699fd7e94aa2/src/shared/modules/search/structured/pathSegmentFilter.js#L84

And unit test case showing an example here: https://github.com/neo-technology/neo4j-insight/blob/6605410f0901eb7210dada6cdb73699fd7e94aa2/src/shared/modules/search/structured/pathSegmentFilter.test.js#L39

Reply:
Hey, thanks for looking at this!

Yeah, we're trying not to be too clever at this point and mark the properties on relationships that aren't on all relationships of a certain type as non-mandatory instead.

Then it's up to the developer to manually filter the typedefs in those cases.

Thanks for those links, looks interesting.